File: examples/python/agent.py
"""
Example agent using any OpenAI-compatible chat completions API.

Usage:
    uv run agent.py <absolute-sandbox-path>

Environment variables:
    LLM_BASE_URL  Base URL of the chat completions API (e.g.: https://api.openai.com)
    LLM_API_KEY   API key
    LLM_MODEL     Model name
"""

import json
import os
import sys
import urllib.request
import urllib.error
from typing import Any

from spadebox import SpadeBox

# --- Configuration ---

BASE_URL = os.environ.get("LLM_BASE_URL", "http://localhost:8324").rstrip("/")
API_KEY = os.environ.get("LLM_API_KEY", "")
MODEL = os.environ.get("LLM_MODEL", "none")

# --- Colors ---

RESET = "\x1b[0m"
BLUE = "\x1b[34m"
GREEN = "\x1b[32m"
RED = "\x1b[31m"
GRAY = "\x1b[90m"
CYAN = "\x1b[36m"

# --- Validate args ---

if len(sys.argv) < 2:
    print("Usage: uv run agent.py <absolute-sandbox-path>", file=sys.stderr)
    sys.exit(1)

sandbox_path = sys.argv[1]

# --- Setup SpadeBox ---

sb = SpadeBox()
sb.enable_files(sandbox_path).enable_js().enable_http().allow("*", ["GET", "HEAD"])

tools = [
    {
        "type": "function",
        "function": {
            "name": t.name,
            "description": t.description,
            "parameters": json.loads(t.input_schema),
        },
    }
    for t in sb.tools()
]

# --- API ---


def chat(messages: list[dict[str, Any]]) -> dict[str, Any]:
    payload = json.dumps(
        {"model": MODEL, "messages": messages, "tools": tools, "tool_choice": "auto"}
    ).encode()
    req = urllib.request.Request(
        f"{BASE_URL}/v1/chat/completions",
        data=payload,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {API_KEY}",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req) as res:
            data = json.loads(res.read())
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"API error {e.code}: {e.read().decode()}") from e
    return data["choices"][0]["message"]


# --- Agent loop ---


def run_turn(messages: list[dict[str, Any]]) -> None:
    while True:
        response = chat(messages)
        messages.append(
            {
                "role": "assistant",
                "content": response.get("content"),
                "tool_calls": response.get("tool_calls"),
            }
        )

        tool_calls = response.get("tool_calls") or []
        if not tool_calls:
            content = response.get("content")
            if content:
                print(f"\n{CYAN}Agent:{RESET} {content}\n")
            return

        for call in tool_calls:
            name = call["function"]["name"]
            args = call["function"]["arguments"]
            print(f"\n{BLUE}[call]{RESET} {GRAY}{name}({args}){RESET}")

            result = sb.call_tool(name, args)
            tag = f"{RED}[error]{RESET}" if result.is_error else f"{GREEN}[ok]{RESET}"
            print(f"{tag} {GRAY}{result.output}{RESET}")

            messages.append(
                {"role": "tool", "tool_call_id": call["id"], "content": result.output}
            )


# --- Main ---

SYSTEM_PROMPT = (
    "You are a helpful agent, help the user and use your tools as appropriate."
)

messages: list[dict[str, Any]] = [{"role": "system", "content": SYSTEM_PROMPT}]

print(f"Agent ready. Sandbox: {sandbox_path}")
print(f"Endpoint: {BASE_URL}, Model: {MODEL}")
print("Type your request, Ctrl+D to exit.\n")

try:
    while True:
        try:
            line = input("> ")
        except EOFError:
            break
        if not line.strip():
            continue
        messages.append({"role": "user", "content": line})
        try:
            run_turn(messages)
        except Exception as e:
            print(f"Error: {e}", file=sys.stderr)
except KeyboardInterrupt:
    pass

File: examples/js/agent.ts
/**
 * Example agent using any OpenAI-compatible chat completions API.
 *
 * Usage:
 *   deno run --allow-read --allow-write --allow-env --allow-ffi --allow-net \
 *            agent.ts <absolute-sandbox-path>
 *
 * Environment variables:
 *   LLM_BASE_URL  Base URL of the chat completions API (e.g.: https://api.openai.com)
 *   LLM_API_KEY   API key
 *   LLM_MODEL     Model name
 */

import process from 'node:process'
import * as readline from 'node:readline'
import { SpadeBox } from '@spadebox/spadebox'

// --- Configuration ---

const BASE_URL = (Deno.env.get('LLM_BASE_URL') ?? 'http://localhost:8324').replace(/\/$/, '')
const API_KEY = Deno.env.get('LLM_API_KEY') ?? ''
const MODEL = Deno.env.get('LLM_MODEL') ?? 'none'

// --- Colors ---

const RESET = '\x1b[0m'
const BLUE = '\x1b[34m'
const GREEN = '\x1b[32m'
const RED = '\x1b[31m'
const GRAY = '\x1b[90m'
const CYAN = '\x1b[36m'

// --- Validate args ---

const sandboxPath = Deno.args[0]
if (!sandboxPath) {
  console.error('Usage: deno run <perms> agent.ts <absolute-sandbox-path>')
  Deno.exit(1)
}

// --- Setup SpadeBox ---

// [snippet: setup]
const sb = new SpadeBox()
sb.enableFiles(sandboxPath)
  .enableJs()
  .enableHttp()
  .allow('*', ['GET', 'HEAD'])
// [/snippet]

// [snippet: tool-definitions]
const tools = sb.tools().map((t) => ({
  type: 'function' as const,
  function: {
    name: t.name,
    description: t.description,
    parameters: JSON.parse(t.inputSchema),
  },
}))
// [/snippet]

// --- Types ---

type Message =
  | { role: 'system'; content: string }
  | { role: 'user'; content: string }
  | { role: 'assistant'; content: string | null; tool_calls?: ToolCall[] }
  | { role: 'tool'; tool_call_id: string; content: string }

interface ToolCall {
  id: string
  type: 'function'
  function: { name: string; arguments: string }
}

interface AssistantMessage {
  role: 'assistant'
  content: string | null
  tool_calls?: ToolCall[]
}

// --- API ---

async function chat(messages: Message[]): Promise<AssistantMessage> {
  const res = await fetch(`${BASE_URL}/v1/chat/completions`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${API_KEY}`,
    },
    body: JSON.stringify({ model: MODEL, messages, tools, tool_choice: 'auto' }),
  })
  if (!res.ok) {
    throw new Error(`API error ${res.status}: ${await res.text()}`)
  }
  const data = await res.json()
  return data.choices[0].message as AssistantMessage
}

// --- Agent loop ---

// [snippet: agent-loop]
async function runTurn(messages: Message[]): Promise<void> {
  while (true) {
    const response = await chat(messages)
    messages.push({ role: 'assistant', content: response.content, tool_calls: response.tool_calls })

    if (!response.tool_calls?.length) {
      if (response.content) console.log(`\n${CYAN}Agent:${RESET} ${response.content}\n`)
      return
    }

    for (const call of response.tool_calls) {
      const { name, arguments: args } = call.function
      console.log(`\n${BLUE}[call]${RESET} ${GRAY}${name}(${args})${RESET}`)

      const result = await sb.callTool(name, args)
      const tag = result.isError ? `${RED}[error]${RESET}` : `${GREEN}[ok]${RESET}`
      console.log(`${tag} ${GRAY}${result.output}${RESET}`)

      messages.push({ role: 'tool', tool_call_id: call.id, content: result.output })
    }
  }
}
// [/snippet]

// --- Main ---

const SYSTEM_PROMPT = `You are a helpful agent, help the user and use your tools as appropriate.`

const messages: Message[] = [{ role: 'system', content: SYSTEM_PROMPT }]

console.log(`Agent ready. Sandbox: ${sandboxPath}`)
console.log(`Endpoint: ${BASE_URL}, Model: ${MODEL}`)
console.log('Type your request, Ctrl+D to exit.\n')

const rl = readline.createInterface({ input: process.stdin, output: process.stdout, prompt: '> ' })
rl.prompt()

for await (const line of rl) {
  if (!line.trim()) continue
  messages.push({ role: 'user', content: line })
  try {
    await runTurn(messages)
  } catch (err) {
    console.error('Error:', err instanceof Error ? err.message : err)
  }
  try { rl.prompt() } catch { /* stdin closed during the last turn */ }
}
